eulerpi/core/data_transformation.py

```diff
@@ -0,0 +1,225 @@
+    ) -> Union[jnp.double, jnp.ndarray]:
+        """Normalize the given data.
+
+        Args:
+            data (Union[jnp.double, jnp.ndarray]): The data to be normalized. Columns correspond to different dimensions. Rows correspond to different observations.
+
+        Returns:
+            Union[jnp.double, jnp.ndarray]: The normalized data.
+        """
+        # possible shapes
+        # normalizing matrix: (d, d)
+        # data: (n, d)
+        # data: (d)
+        # The correct output shape is (n, d) or (d) depending on the input shape.
+        return jnp.inner(data - self.mean_vector, self.normalizing_matrix)
+
+    def jacobian(self, data: jnp.ndarray) -> jnp.ndarray:
+        return self.normalizing_matrix
+
+    def _tree_flatten(self):
+        """This function is used by JAX to flatten the object for JIT compilation."""
+        children = (
+            self.normalizing_matrix,
+            self.mean_vector,
+        )  # arrays / dynamic values
+        aux_data = {}  # static values
+
+        return (children, aux_data)
+
+    @classmethod
+    def _tree_unflatten(cls, aux_data, children):
+        """This function is used by JAX to unflatten the object for JIT compilation."""
+        return cls(*children, **aux_data)
+
+
+tree_util.register_pytree_node(
+    DataNormalizer,
+    DataNormalizer._tree_flatten,
+    DataNormalizer._tree_unflatten,
+)
+
+
+class DataPCA(DataTransformation):
+    """The DataPCA class can be used to transform the data using the Principal Component Analysis."""
+
+    def __init__(
+        self,
+        pca: Optional[PCA] = None,
+    ):
+        """Initialize a DataPCA object.
+
+        Args:
+            pca (Optional[PCA], optional): The PCA object to be used for the transformation. Defaults to None.
+        """
+        super().__init__()
+
+        self.pca = pca
+
+    @classmethod
+    def from_data(
+        cls, data: jnp.ndarray, n_components: Optional[int] = None
+    ) -> "DataTransformation":
+        """Initialize a DataPCA object by calculating the PCA from the given data.
+
+        Args:
+            data (jnp.ndarray): The data to be used for the PCA.
+            n_components (Optional[int], optional): The number of components to keep. If None is passed, min(n_samples,n_features) is used. Defaults to None.
+
+        Returns:
+            DataTransformation: The initialized DataPCA object.
+        """
+        instance = cls()
+        instance.pca = PCA(n_components=n_components)
+        instance.pca.fit(data)
+        return instance
+
+    def transform(self, data: jnp.ndarray) -> jnp.ndarray:
+        """Transform the given data using the PCA.
+
+        Args:
+            data (jnp.ndarray): The data to be transformed.
+
+        Returns:
+            jnp.ndarray: The data projected onto and expressed in the basis of the principal components.
+        """
+        return self.pca.transform(data.reshape(-1, data.shape[-1])).reshape(
```

Comment:
We might want to change the orientation of the data internally in the future.

Reply:
Yep, also using basic numpy functions like matmul instead of inner would be amazing.